Add stdlib::get_from_bucket functionality and tests

jeffmccune opened the pull request:
Test coverage for the following functions which are optionally included via
feature flags.

 * stdlib::init_gsutil_crcmod_el
 * stdlib::get_from_bucket

# Do not merge until

 * [x] gsutil related integration tests have been added
 * [x] #12 merged
 * [x] #14 merged

jeffmccune committed:
(maint) Remove TODO comments

They're tracked in the bug tracker now.

jeffmccune committed:
Add stdlib::mandatory_argument

Helper function to log missing mandatory arguments in a unified way.

jeffmccune committed:
Add stdlib::get_from_bucket function

Behavior to download a file from a storage bucket with retry behavior to
handle transient failures.

jeffmccune committed:
Add enable_get_from_bucket feature flag

jeffmccune committed:
Add kitchen-terraform inspec tests for gsutil functionality

Test coverage for the following functions which are optionally included
via feature flags.

 * stdlib::init_gsutil_crcmod_el
 * stdlib::get_from_bucket

In addition:

 * stdlib::cmd

jeffmccune commented on examples/gsutil/main.tf:
```diff
@@ -0,0 +1,93 @@
+  storage_class = "REGIONAL"
+}
+
+resource "google_storage_bucket_object" "message" {
```
Left off here.  When running kitchen converge from inside `make integration_test_shell` and with `source test/ci_integration.sh && setup_environment`

```
       Error: Error applying plan:

       1 error(s) occurred:

       * module.example.google_storage_bucket.example: 1 error(s) occurred:

       * google_storage_bucket.example: googleapi: Error 403: ci-startup-scripts@startup-scripts.iam.gserviceaccount.com does not have storage.buckets.create access to project 1010336466588., forbidden

       Terraform does not automatically rollback in the face of errors.
       Instead, your Terraform state file has been partially updated with
       any resources that successfully completed. Please address the error
       above and apply again to incrementally change your infrastructure.


>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Converge failed on instance <gsutil-local>.  Please see .kitchen/logs/gsutil-local.log for more details
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
```

jeffmccune commented:
@adrienthebo This is ready for review.  It's based off #14

jeffmccune committed:
Remove needless integration test

The removed example doesn't test anything related to startup scripts as
mentioned in the review comment
https://github.com/terraform-google-modules/terraform-google-startup-scripts/pull/14#discussion_r256599608

adrienthebo commented on examples/gsutil/outputs.tf:
```diff
@@ -0,0 +1,28 @@
+ * limitations under the License.
+ */
+
+output "project_id" {
```
Can we add descriptions to `project_id` and `region`?

jeffmccune replied:
```diff
commit f83b4e6904218757c46a0f02cdb3421f02c5a551
Author: Jeff McCune <jeff@openinfrastructure.co>
Date:   Fri Feb 15 13:26:30 2019 -0800

    Fixup output descriptions

diff --git a/examples/gsutil/outputs.tf b/examples/gsutil/outputs.tf
index 0438282..de7cd18 100644
--- a/examples/gsutil/outputs.tf
+++ b/examples/gsutil/outputs.tf
@@ -15,11 +15,13 @@
  */
 
 output "project_id" {
-  value = "${var.project_id}"
+  description = "The project id used when managing resources."
+  value       = "${var.project_id}"
 }
 
 output "region" {
-  value = "${var.region}"
+  description = "The region used when managing resources."
+  value       = "${var.region}"
 }
 
 output "nat_ip" {

```

adrienthebo replied:
@jeffmccune I'm not seeing that commit on this PR, could you make sure that you've pushed?

adrienthebo commented on variables.tf:
```diff
@@ -17,3 +17,8 @@ variable "enable_init_gsutil_crcmod_el" {
+
+variable "enable_get_from_bucket" {
+  description = "If not false, include stdlib::get_from_bucket() prior to executing startup-script-custom.  Requires gsutil in the PATH.  See also enable_init_gsutil_crcmod_el feature flag."
+  default     = "false"
```
Out of curiosity, why is this feature opt-in? This seems like a handy thing to have, what's the use case for someone wanting to have this disabled?

jeffmccune replied:
See #12 which was added based on input from Morgante.  The overall design philosophy is that only truly platform-independent things should be always-enabled.  Today that's logging and config loading behavior.

Platform specific behavior is to be opt-in to make it easier to maintain and support the module.  In this case get_from_bucket depends on gsutil, which depends on platform specific compilation of the crc32 library [1].

To limit scope this is only implemented today for EL systems.  The debian family will need to opt-out of EL behavior and opt-in to their own behavior.

[1] https://cloud.google.com/storage/docs/gsutil/addlhelp/CRC32CandInstallingcrcmod

adrienthebo commented on test/ci_integration.sh:
```diff
@@ -46,6 +47,10 @@ setup_environment() {
+
+  # Unique ID used for resource names
+  uuid="$(</proc/sys/kernel/random/uuid)"
+  export UNIQUE_ID="${uuid:0:8}"
```
Where is this being used? Grepping through the project source doesn't show anything.

jeffmccune replied:
Good catch, removed.

```diff
commit b186b3ab9cd01f742c3f329cb3b4d16db5a9ea12
Author: Jeff McCune <jeff@openinfrastructure.co>
Date:   Fri Feb 15 13:33:25 2019 -0800

    Fixup remove unused UNIQUE_ID var

diff --git a/test/ci_integration.sh b/test/ci_integration.sh
index 0ef3f4a..4f970a8 100755
--- a/test/ci_integration.sh
+++ b/test/ci_integration.sh
@@ -34,7 +34,6 @@ finish() {
 # use with kitchen-terraform, inspec, and gcloud.
 setup_environment() {
   local tmpfile
-  local uuid
   tmpfile="$(mktemp)"
   echo "${SERVICE_ACCOUNT_JSON}" > "${tmpfile}"
 
@@ -47,10 +46,6 @@ setup_environment() {
   # Terraform input variables
   export TF_VAR_project_id="${PROJECT_ID}"
   export TF_VAR_region="${REGION:-us-east4}"
-
-  # Unique ID used for resource names
-  uuid="$(</proc/sys/kernel/random/uuid)"
-  export UNIQUE_ID="${uuid:0:8}"
 }
 
 main() {
```

adrienthebo commented on examples/gsutil/variables.tf:
```diff
@@ -0,0 +1,33 @@
+  description = "The region to deploy to"
+}
+
+variable "url" {
```
Is this variable in use for this example?

jeffmccune replied:
No, good catch.  Removed.

```diff
commit ee3929913ec3bfe800fed5c4c5af05f15f1b735f
Author: Jeff McCune <jeff@openinfrastructure.co>
Date:   Fri Feb 15 14:39:04 2019 -0800

    Fixup Remove unused URL variable

diff --git a/examples/gsutil/variables.tf b/examples/gsutil/variables.tf
index d3d341c..4cca17e 100644
--- a/examples/gsutil/variables.tf
+++ b/examples/gsutil/variables.tf
@@ -22,11 +22,6 @@ variable "region" {
   description = "The region to deploy to"
 }
 
-variable "url" {
-  description = "The url to fetch in the startup script.  This URL is passed via instance metadata to the startup script.  (e.g. ifconfig.co/city)"
-  default     = "http://ifconfig.co/json"
-}
-
 variable "message" {
   description = "The content to place in a bucket object message.txt. startup-script-custom fetches this object and validate this message against the content as an end-to-end example of stdlib::get_from_bucket()."
   default     = "Hello World! uuid=0afce28a-057b-42cf-a90f-493de3c0666b"
```

adrienthebo commented on examples/gsutil/main.tf:
```diff
@@ -0,0 +1,94 @@
+  byte_length = 4
+}
+
+#    webhook_token = "${random_id.startup_scripts_github_webhook_token.hex}"
```
Can we drop this comment?

jeffmccune replied:
Sure thing

```diff
commit 3a5d40f3ecd15a3828259619329a6d7830fa82bf
Author: Jeff McCune <jeff@openinfrastructure.co>
Date:   Fri Feb 15 14:39:52 2019 -0800

    Fixup drop webhook_token comment

diff --git a/examples/gsutil/main.tf b/examples/gsutil/main.tf
index 60d7a9b..ac574b4 100644
--- a/examples/gsutil/main.tf
+++ b/examples/gsutil/main.tf
@@ -36,8 +36,6 @@ resource "random_id" "resource_name_suffix" {
   byte_length = 4
 }
 
-#    webhook_token = "${random_id.startup_scripts_github_webhook_token.hex}"
-
 # Storage bucket used by startup-script-custom and stdlib::get_from_bucket
 resource "google_storage_bucket" "example" {
   name          = "startup-scripts-${random_id.resource_name_suffix.hex}"
```

adrienthebo commented on files/get_from_bucket.sh:
```diff
@@ -0,0 +1,60 @@
+# Given a url and filename, download an object to the vardir. This function uses
+# gsutil to fetch from the bucket.  Note, the service account for the instance
+# must be properly configured with a role having authorization to get objects
+# from the bucket.  See Product/mktmail/commit/39fa9b8 for an example of the
```
What is `Product/mktmail/commit/39fa9b8` ?

jeffmccune replied:
Cruft, removed it in 513dc7b.  Good catch.

adrienthebo commented on test/integration/simple_example/controls/gcloud.rb:
```diff
@@ -13,13 +13,16 @@
   title "With the simple example of startup-script-custom calling stdlib::info and stdlib::cmd"
 
-  describe command("gcloud compute instances list --project #{project_id}") do
+  describe command("gcloud compute instances get-serial-port-output startup-scripts-example1 --project #{project_id} --zone #{zone}") do
```
This is an awesome test, 💯

jeffmccune replied:
Thanks!

adrienthebo commented on examples/gsutil/templates/startup-script-custom.tpl:
```diff
@@ -0,0 +1,16 @@
+#
+# Check for crcmod
+# https://cloud.google.com/storage/docs/gsutil/addlhelp/CRC32CandInstallingcrcmod
+stdlib::cmd gsutil version -l
```
What is this doing? Do we have a test assertion checking for the output from this command?

jeffmccune replied:
Good catch, this touches on incomplete tests that were half written.

I've started cleaning up the test cases and will continue, but probably won't finish it before the weekend.

Tests to add without any other cruft:

 * [x] Expect gsutil version -l to report `compiled crcmod: False` prior to calling stdlib::init_gsutil_crcmod_el
 * [x] Expect `compiled crcmod: True` after calling stdlib::init_gsutil_crcmod_el
 * [x] Expect stdlib::get_from_bucket to download message.txt when the instance has `storage.objects.get` access.
 * [x] Expect stdlib::get_from_bucket to retry up to 10 times when the instance does not have `storage.objects.get` access

adrienthebo commented on test/integration/gsutil/controls/gcloud.rb:
```diff
@@ -0,0 +1,34 @@
+    its('stdout') { should match(/startup-scripts-example.*RUNNING/) }
+  end
+
+  describe command("gcloud compute instances get-serial-port-output startup-scripts-example1 --project #{project_id} --zone #{zone}") do
```
Blocker: this is testing the wrong VM.

jeffmccune replied:
Great catch, fixed in the WIP test clarity improvements.

adrienthebo commented on test/integration/gsutil/controls/gcloud.rb:
```diff
@@ -0,0 +1,34 @@
+  describe command("gcloud compute instances get-serial-port-output startup-scripts-example1 --project #{project_id} --zone #{zone}") do
+    its('exit_status') { should be 0 }
+    its('stderr') { should match(%r{Specify --start=\d+ in the next get-serial-port-output invocation to get only the new output starting from here})}
+    its('stdout') { should match(%r{Info \[\d+\]: Fetching http://ifconfig\.co/json}) }
```
Blocker: this isn't testing the gsutil behavior

jeffmccune replied:
Great catch, fixed in the WIP test clarity improvements.

adrienthebo reviewed (dismissed):
# Review summary

Generally looks good, there's one bug we need to chase down and a few cosmetic changes. Let's fix those up and get this merged. Thanks!

## Blockers

- [x] Update gsutil integration tests to target the `gsutil` host
- [x] Update gsutil integration tests to verify the `get_from_bucket` function

## Best practices

- [x] Remove dead code comment in `examples/gsutil/main.tf`
- [x] Remove dead `url` variable in `examples/gsutil/variables.tf` (if it's truly dead)
- [x] Document examples/gsutil outputs

## Review

- [x] Clarify comment in files/get_from_bucket.sh
- [x] Clarify the use of the `UNIQUE_ID` environment variable set in `test/ci_integration.sh`

jeffmccune committed:
Fixup output descriptions

jeffmccune committed:
Fixup remove unused UNIQUE_ID var

jeffmccune committed:
Fixup Remove unused URL variable

jeffmccune committed:
Fixup drop webhook_token comment

jeffmccune committed:
Fixup scrub comment

jeffmccune committed:
Fixup gsutil enable_init_gsutil_crcmod_el test

jeffmccune committed:
Fixup tests.  Clearly exercise the behavior being implemented and no more.

 * [x] Expect gsutil version -l to report `compiled crcmod: False` prior to calling stdlib::init_gsutil_crcmod_el
 * [x] Expect `compiled crcmod: True` after calling stdlib::init_gsutil_crcmod_el
 * [ ] Expect stdlib::get_from_bucket to download message.txt when the instance has `storage.objects.get` access.
 * [x] Expect stdlib::get_from_bucket to retry up to 10 times when the instance does not have `storage.objects.get` access

jeffmccune committed:
Fixup tests.  Clearly exercise the behavior being implemented and no more.

 * [x] Expect gsutil version -l to report `compiled crcmod: False` prior to calling stdlib::init_gsutil_crcmod_el
 * [x] Expect `compiled crcmod: True` after calling stdlib::init_gsutil_crcmod_el
 * [x] Expect stdlib::get_from_bucket to download message.txt when the instance has `storage.objects.get` access.
 * [ ] Expect stdlib::get_from_bucket to retry up to 10 times when the instance does not have `storage.objects.get` access

jeffmccune commented:
Next steps on this are to create a gsutil2 instance which is a copy of the current gsutil1 instance. 

 * gsutil2 will have storage.objects.get access for the positive test
 * gsutil1 will retain the current behavior to include a negative test of retry behavior when access is denied.

adrienthebo commented:
@jeffmccune we're getting CI failures, could you take a look? Also, it looks like we're getting some odd formatting in CI, you may want to disable color in the terraform verifier like what we're doing for the [project factory .kitchen.yaml](https://github.com/terraform-google-modules/terraform-google-project-factory/blob/master/.kitchen.yml#L24).

jeffmccune commented:
Will do, let's leave this on hold until I clean up the failures.

jeffmccune committed:
(maint) Disable kitchen color output

Because CI colors are flashing which make reading the job logs very difficult.

jeffmccune committed:
(WIP) Fixup race condition collecting serial console output

Without this patch the instance startup sequence may not be completed when the
test suite collects gcloud console output for testing.  This creates a race
condition.

This patch addresses the problem by retrying the gcloud command until the
marker `systemd: Startup finished` is found in the output.

jeffmccune committed:
Fixup - Associate a service account with the gsutil instance

Without this patch the instance used for positive behavior tests of the gsutil
command does not have read access to the bucket fixture.  This causes failures.

This patch fixes the problem by associating the CI service account with the
instance.  A future improvement could introduce a new SA with more limited
access, e.g. get only.  Negative tests to cover retry behavior may also be
introduced in a future improvement.

jeffmccune commented on test/integration/gsutil/controls/gcloud.rb:
```diff
@@ -0,0 +1,72 @@
+        @cmd = command("#{get_serial_port_output} --project #{attribute('project_id')} --zone #{attribute('region')}-a")
+        if not %r{systemd: Startup finished}.match(@cmd.stdout)
+          raise StandardError, "Not found: 'systemd: Startup finished' in console output, cannot proceed"
+        end
```
@adrienthebo Spent some time this evening tracking down the test failures.  One of two issues was a race condition with the gcloud command being executed while startup was in-progress.  I'm curious on your feedback regarding this `Retriable` approach to solving the issue.  I generally avoid `before :all`, but it seems prudent in this situation.

adrienthebo replied:
This does work, but waiting for a startup script to finish running feels like more of a provisioning concern rather than a testing issue. I'd expect for `kitchen converge` to block until the VMs are up and ready and that `kitchen verify` can be run. I've implemented this for the Forseti module with similar logic at https://github.com/terraform-google-modules/terraform-google-forseti/blob/master/test/fixtures/simple_example/main.tf#L57-L71 - could we shift this logic to a `null_resource` with a `local-exec` provisioner?

jeffmccune commented on examples/gsutil/main.tf:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  service_account {
+    email  = "${var.service_account_email}"
```
@adrienthebo The second of two issues with the test failures was the instance didn't have `storage.buckets.get` permission and as such the stdlib::get_from_bucket was failing.  I re-use the SA email address all the way from the top level `SERVICE_ACCOUNT_JSON` input environment variable and thread it through the `setup_environment` function in ci_integration.sh.

I plan to setup a negative test to validate the retry logic of get_from_bucket, but this seems like it should be a future PR to keep this one from growing too much larger.  Thoughts?

adrienthebo replied:
I didn't see that CI failure; when I looked I saw that a GCE instance startup logs were dumped to the CI logs and had trouble investigating further. What you're describing makes sense, and since CI is green this looks resolved to me.

👍 for adding the negative test in a new PR.

jeffmccune commented:
@adrienthebo I fixed up the tests.  I've left the sequence of patches in their "fixup" commits to make review easier.  I'd like to avoid adding much more code to this PR to prevent it from growing larger than it already is.  Could you please review again, putting aside git hygiene.  If approved, I'll rebase ahead of merge.  What do you think?

Regarding [my comment](https://github.com/terraform-google-modules/terraform-google-startup-scripts/pull/15#issuecomment-464258735), "Next steps on this are to create a gsutil2 instance which is a copy of the current gsutil1 instance." I propose that be moved to a future, smaller, PR.

adrienthebo reviewed (commented):
One suggestion that we move the startup-script delay/retry logic from the verification step to the provisioning step.

@jeffmccune since this works do you want to revise the retry logic in a future PR, or do you want to give it a shot here? Ping me via chat if you'd like to chat a bit more about this.

jeffmccune commented:
If you could lead me along the path with a link to prior art, I'm happy to
do it here as part of the rebase work that needs to be done.  Otherwise
it's probably best to do this in a separate PR as I'd be covering new (to
me) ground moving the logic to the converge / create steps.

morgante commented:
I vote in favor of merging now and adding the retry logic in a follow-on PR.
